File: Blind75/Graphs/Alien Dictionary.py

#Topological Sort Kahn's algo BFS
from collections import deque


class Solution:
    def foreignDictionary(self, words: List[str]) -> str:
        # Step 1: Initialize adjacency and indegree maps
        adj = {c: set() for w in words for c in w}
        indegree = {c: 0 for c in adj}

        # Step 2: Build the graph by comparing adjacent words
        for i in range(len(words) - 1):
            w1, w2 = words[i], words[i + 1]
            minLen = min(len(w1), len(w2))

            # Invalid prefix case
            if len(w1) > len(w2) and w1[:minLen] == w2[:minLen]:
                return ""

            # Find the first differing character
            for j in range(minLen):
                if w1[j] != w2[j]:
                    if w2[j] not in adj[w1[j]]:
                        adj[w1[j]].add(w2[j])
                        indegree[w2[j]] += 1
                    break

        # Step 3: Topological sort using Kahn’s Algorithm (BFS)
        q = deque([c for c in indegree if indegree[c] == 0])
        res = []

        while q:
            char = q.popleft()
            res.append(char)
            for nei in adj[char]:
                indegree[nei] -= 1
                if indegree[nei] == 0:
                    q.append(nei)

        # Step 4: Check for cycles
        if len(res) != len(indegree):
            return ""

        return "".join(res)

File: NeedCode 150 (JAVA)/Graphs/Alien Dictionary.java
import java.util.*;

class Solution {
    public String alienOrder(String[] words) {

        // Character -> characters that must come after it
        Map<Character, Set<Character>> graph = new HashMap<>();

        // Character -> number of incoming edges
        Map<Character, Integer> indegree = new HashMap<>();

        // Initialize every character
        for (String word : words) {
            for (char c : word.toCharArray()) {
                graph.putIfAbsent(c, new HashSet<>());
                indegree.putIfAbsent(c, 0);
            }
        }

        // Build graph from adjacent words
        for (int i = 0; i < words.length - 1; i++) {

            String word1 = words[i];
            String word2 = words[i + 1];

            int minLen = Math.min(word1.length(), word2.length());

            boolean foundDifference = false;

            for (int j = 0; j < minLen; j++) {

                char c1 = word1.charAt(j);
                char c2 = word2.charAt(j);

                if (c1 != c2) {

                    // c1 must come before c2
                    // Add edge only if it doesn't already exist
                    if (graph.get(c1).add(c2)) {
                        indegree.put(c2, indegree.get(c2) + 1);
                    }

                    foundDifference = true;
                    break;
                }
            }

            // Invalid case:
            // ["abc", "ab"]
            if (!foundDifference && word1.length() > word2.length()) {
                return "";
            }
        }

        // Topological sort using Kahn's algorithm
        Queue<Character> queue = new LinkedList<>();

        for (char c : indegree.keySet()) {
            if (indegree.get(c) == 0) {
                queue.offer(c);
            }
        }

        StringBuilder result = new StringBuilder();

        while (!queue.isEmpty()) {

            char curr = queue.poll();
            result.append(curr);

            for (char next : graph.get(curr)) {

                indegree.put(next, indegree.get(next) - 1);

                if (indegree.get(next) == 0) {
                    queue.offer(next);
                }
            }
        }

        // If not all characters were processed, there is a cycle
        if (result.length() != indegree.size()) {
            return "";
        }

        return result.toString();
    }
}
